### PR TITLE
feat: company-history.mjs — employer silence & repost facts (#1711)

### LINKED ISSUE #1711: feat: company-history.mjs — per-company employer-silence & repost evidence cards
## Problem

A candidate about to spend an evening tailoring a CV often already has the decisive evidence in their own files — unjoined:

- the tracker says they applied to this company in March and never heard back,
- follow-ups.md says two nudges went unanswered,
- scan-history.tsv says this exact role has been re-listed several times in 90 days.

Nothing joins these per company today. The scale of the underlying problem is well documented: 53% of job seekers report being ghosted by an employer in the past year (iHire 2026, up from 38% in 2024); 81% of recruiters admit their employer posts ghost jobs (MyPerfectResume); roughly 1 in 3 postings has no current intent to hire behind it. SaaS job boards can't solve this for the user — they don't own the user's history. career-ops does.

## Proposal: `company-history.mjs` — read-only evidence cards, two fact axes

A sibling aggregator in the `detect-reposts.mjs` / `process-quality.mjs` family (JSON default, `--summary`, `--self-test`, zero writes, zero new data files). It joins tracker + follow-ups + scan-history (+ a status-log source via defensive dynamic import, for when funnel infrastructure lands) on the normalized company key, and emits one card per company with **two independent axes that never merge into a tier**:

- **`postingChurn`** — posting-level, pre-application: `reposts-detected | none-detected | no-scan-data`, from detect-reposts clusters (count, span, last-seen). Action space: how much to trust the posting.
- **`responsiveness`** — relationship-level, post-application: `responded-before | silent-on-you | mixed | no-history`, from the user's own Applied/Rejected/Interview rows. A rejection counts as a response. Action space: how much effort to keep investing.

These are different phenomena with different causal stories, so they are reported as two labeled axes; each surface consumes only the one that's relevant to it.

## Honesty rules (designed in, asserted by self-test)

- **No verdict vocabulary.** Output is descriptive past-tense fact ("silent 121d since 2026-03-10, 1 follow-up unanswered"), never "ghosted"-as-verdict, never "risk"/score/tier. The self-test asserts the rendered summary contains neither `/risk/i` nor `/ghost/i`.
- **Hygiene before evidence.** The dominant generator of "silent" rows is the user's own stale tracker. The summary leads with an aged-Applied nudge list, and every silent fact carries the exact `set-status.mjs` command that clears it if the company actually did respond.
- **Right-censoring.** Rows younger than the silence window (default 28d = 2× the typical 5–14d first-response upper bound, `--silence-window` configurable) are pending, never silent — and the summary prints once that many real processes exceed the window.
- **Follow-up compliance is never punished.** Unanswered follow-ups upgrade a silent fact's *confidence* annotation only; they never change the label.
- **Evidence lifecycle.** Every fact is dated; facts older than 365d are stale and excluded from labels unless `--include-stale`; a later response from the same company supersedes in the headline (`mixed`) with history retained — which keeps the output falsifiable.
- **Join guard.** Company names that normalize to an empty key (non-Latin collapse) are excluded from joins and counted in `dataQuality.unjoinable` — never merged into another company's card.

## Surfaces

- **oferta Block G signal #4 (reposting)** — upgraded to read the `postingChurn` clusters from `--company "{company}"` instead of a raw scan-history grep. Churn axis only, same reliability rating, same tier mechanics — this stays inside the existing signal, so no double-counting.
- **oferta: prior-contact FYI (non-scoring)** — when the responsiveness label is `silent-on-you`/`mixed`, one informational line in the report, explicitly barred from the 1–5 score and the legitimacy tier (personal response history says nothing about whether a *new posting* is real).
- **patterns** — Company History lens: hygiene nudge first, cards silent-first, mandatory causal-humility framing (facts about the user's own files, innocent explanations, no verdicts).
- **followup** — expectation-setting for silent-on-you companies (cap further investment; sending stays the user's call).
- **tracker** — one summary line offering the lens when any silent-on-you company exists.

## Relationship to existing work

- **Block G signal #4** established the precedent that repost data informs posting legitimacy — this deepens that signal without adding a new one.
- **#1506 (shared behavioral-signal layer)** is the natural future consumer: everything here is company-attributable and candidate-anonymous by construction; day-level dates stay local (month-trim would happen at any shared-layer boundary).
- The per-company join spine is deliberately reusable — the same card structure can later carry other per-company lenses (process-quality friction, via-channel outcomes, ATS vendor behavior) toward a fuller "company dossier" view.

## Scope guardrails (v1)

No writes of any kind, no prediction, no composite score, no cross-install sharing (that's #1506 itself), no interview-stage ghosting (process-quality's axis). Ships from main with no dependency on unmerged funnel PRs — the status-log source degrades cleanly until that infrastructure exists.

PR with the implementation (script + self-test + external suite + `parseFollowups(content)` export refactor + registration + docs + mode wiring) to follow.

### PR DESCRIPTION
**Pull request by Scott-Emberson**
Closes #1711.

## What

`company-history.mjs` — a read-only, per-company evidence-card aggregator in the `detect-reposts.mjs` / `process-quality.mjs` family. It joins the tracker, `data/follow-ups.md`, and `data/scan-history.tsv` on the normalized company key (empty keys excluded and counted in `dataQuality.unjoinable`, never merged into another company's card) and emits one card per company with two independent fact axes:

- **`postingChurn`** — `reposts-detected | none-detected | no-scan-data`, from detect-reposts clusters (repostCount / daysSpan / lastSeen), re-keyed through `normalizeCompany`.
- **`responsiveness`** — `responded-before | silent-on-you | mixed | no-history`, from the user's own tracker rows. A rejection counts as a response. Applied dates resolve status-log → notes (`Applied YYYY-MM-DD`) → date column (annotated `dateBasis: "evaluation-date"` as weakest evidence).

CLI: JSON default, `--summary` (hygiene nudge first, cards silent-first, window caveat printed once), `--company <name>`, `--silence-window N` (default `templates/benchmarks.yml` when present, else 28), `--include-stale`, `--self-test`; unrecognized flags rejected with usage (#1635 convention). A `status-log` source loads only via a try/catch'd dynamic `import('./funnel-velocity.mjs')`, so this PR ships from main with **no dependency on the open funnel PRs** — `medianResponseDays` lights up when that lands.

## Honesty rules (enforced, not aspirational)

- **Facts, not verdicts.** Output is descriptive past-tense ("silent 121d since 2026-03-10, 1 follow-up, confirmed-by-followups"), never "ghosted"/"risk"/score/tier. The self-test asserts the rendered summary matches neither `/risk/i` nor `/ghost/i`.
- **Hygiene before evidence.** Aged-Applied nudge list leads the summary; every silent fact carries the exact `set-status.mjs` command (real flag syntax — `--note`) that clears it if the response just wasn't logged.
- **Right-censoring.** Rows younger than the silence window are pending, never silent.
- **Follow-up compliance never punished.** Follow-ups affect only the `confidence` annotation, never the label.
- **Evidence lifecycle.** Dated facts; >365d = stale, excluded from labels unless `--include-stale`; later response supersedes into `mixed` with history retained.
- Innocent-explanations line on every card with a silent fact.

## Mode wiring

- **oferta Block G signal #4** upgraded: read `postingChurn` clusters from `--company "{company}"` — churn axis ONLY; responsiveness facts are explicitly barred from the signal and the legitimacy tier (personal history says nothing about whether a new posting is real, and keeping it inside the existing signal avoids double-counting).
- **oferta**: new non-scoring "Prior-contact FYI" subsection — one informational line when the label is `silent-on-you`/`mixed`, barred from the 1–5 score and the Assessment tier.
- **patterns**: Company History lens (hygiene first, silent-first, mandatory causal-humility framing).
- **followup**: expectation-setting for silent-on-you companies; the follow-up itself is never skipped or downgraded.
- **tracker**: offers `--summary` when any silent-on-you company exists.
- **_shared.md** signals table: reposting row now sourced via company-history.mjs.

## Also in this PR

- `followup-cadence.mjs`: `parseFollowups(content)` extracted and exported (internal callers use a thin `readFollowups()` wrapper — byte-identical behavior) so the follow-ups table parse isn't replicated. Pin-directive lines are not sent follow-ups; covered by new tests.
- Registration: test-all runs `company-history.mjs --self-test` + `company-history.test.mjs`; SYSTEM_PATHS covers both files (`validate-system-paths-coverage` green); `docs/SCRIPTS.md` entry documents both axes, flags, exit codes, and the no-verdict vocabulary contract; five test-all mode-doc assertions lock the wiring.

## Tests

- `company-history.mjs --self-test`: **38 checks** — join fixtures (case/punct variants, empty-key → unjoinable), label goldens incl. right-censoring / rejection-is-an-answer / mixed supersedence / stale + `--include-stale`, pin-line exclusion, confidence-not-label, per-source absent-file degradation, absent funnel-velocity module, `--company` lookup, a distribution-sanity fixture (not everything labels silent), and the vocabulary assertions.
- `company-history.test.mjs`: **72 assertions** — 3-company multi-source JSON-shape scenario, silence-window boundary goldens (window-1 pending / at-window and window+1 silent), dateBasis provenance priority, real-file loader round-trips, CLI smoke tests.
- `followup-cadence.test.mjs`: 10 passed (6 new for the export).
- Full suite: `node test-all.mjs` → **1542 passed, 0 failed**.
- E2E sandbox run (fabricated tracker / follow-ups / scan-history via `CAREER_OPS_TRACKER` + path overrides) verified both axes, the hygiene nudge, the pin-line exclusion, and zero writes (`git status` clean after every run).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a read-only `company-history` CLI that aggregates per-company evidence cards, including `--summary`, optional company lookup, and JSON output.
  * Added labeled evidence axes for responsiveness (silence-window + stale handling) and posting churn (repost clusters / no-scan-data).
* **Bug Fixes**
  * Improved resilience when optional history sources are missing or partially available.
  * Strengthened `--silence-window` validation and edge-case behavior.
* **Documentation**
  * Updated multiple mode instructions to reference the new evidence cards and “repost-churn” lens.
* **Tests**
  * Added an external test runner for `company-history`, plus expanded overall test execution.

### CODE REVIEW

File: company-history.mjs [hunk 1]
```diff
@@ -0,0 +1,787 @@
...
+  } else {
+    const run = async () => {
+      const tracker = loadTrackerRows(CAREER_OPS);
+      const followups = loadFollowupRows(CAREER_OPS, followupsOverride);
+      const scanHistory = loadRepostClusters(CAREER_OPS, scanHistoryOverride);
+      const statusLog = await loadStatusLogSource();
+
+      const silenceWindowDays = silenceWindowArg !== undefined && Number.isFinite(parseInt(silenceWindowArg, 10))
+        ? parseInt(silenceWindowArg, 10)
+        : resolveDefaultSilenceWindow(CAREER_OPS);
```
[the remark this answers is not in the document]
Comment (Scott-Emberson) [reply company-history.mjs]: Fixed in 5636f6d — `--silence-window` is now validated in parseArgs as a positive integer (`/^[1-9]\d*$/`): non-numeric, zero, and negative values fail fast with a clear error + usage, before any label computation. The silent NaN fallback in the main path is gone. 4 new CLI test cases cover abc / 0 / -5 / valid 21.

File: modes/oferta.md [hunk 2]
```diff
@@ -201,6 +202,14 @@ This signal does not change the High Confidence / Proceed with Caution / Suspici
...
 **Context Notes:** Any caveats (niche role, government job, evergreen position, etc.) that explain potentially concerning signals.
 
+### Prior-contact FYI (non-scoring)
+
+Check the `responsiveness` axis of the same `node company-history.mjs --company "{company}"` card. If `responsiveness.label` is `silent-on-you` or `mixed`, append ONE informational line to the report:
+
+> Note: you applied to {company} on {date}; no response in {N}d after {M} follow-ups. Not a legitimacy signal — factor into how much effort to invest.
+
+This is information about **your own history** with the company, not about this posting. It must NOT alter the 1-5 score and must NOT alter the Assessment tier above — those are driven exclusively by the `postingChurn` axis and the other Block G signals. If the label is `responded-before` or `no-history`, say nothing (silence is fine; no note needed).
+
```
[the remark this answers is not in the document]
Comment (Scott-Emberson) [reply modes/oferta.md:355]: Fixed in 5636f6d — the FYI now branches on the label: `silent-on-you` keeps the no-response wording; `mixed` gets its own message naming both the response and the silence. Still informational-only, still barred from the score and Assessment tier.

File: test-all.mjs [hunk 3]
```diff
@@ -1368,6 +1370,42 @@ if ((batchPromptDoc.match(/advertised_comp/g) || []).length >= 2) {
...
+} else {
+  fail('tracker mode missing company-history.mjs reference');
+}
+
+if (shared.includes('company-history.mjs')) {
+  pass('_shared.md signals table joins the reposting signal through company-history.mjs');
+} else {
+  fail('_shared.md signals table missing company-history.mjs join reference');
+}
+
```
[the remark this answers is not in the document]
Comment (Scott-Emberson) [reply test-all.mjs:2489]: Fixed in 5636f6d — followup assertion now requires BOTH `company-history.mjs` AND `silent-on-you` (was OR); patterns assertion verifies `hygiene` appears BEFORE the `aged-Applied` reference via index comparison, pinning the documented precedence, not just the mentions.

**Review by santifer — CHANGES_REQUESTED** [review]
Third in the block order (#1694 → #1695 → **#1712** → #1719 → #1714 — anchor on #1694). Full isolated review: merge-sim against today's main is CLEAN, suite 1632/0, `company-history --self-test` 38/38, its own test file 76/76, §55 contract-freeze green.

**What holds up well:** this is how to touch Block G — the `_shared.md` change re-routes the *same* reposting signal through the script without moving weights, thresholds or reliability, and the new oferta subsection is explicitly non-scoring with a test-all assertion ("Not a legitimacy signal") freezing that promise in CI. Reuse over reimplementation throughout: `detectReposts` from #1212, `tracker-parse`, `tracker-utils`, `followup-cadence` (whose `parseFollowups` refactor keeps behavior, with regression tests). Read-only, no new data files.

**The one real ask — the two speculative hooks:** this PR pre-wires (a) a dynamic import of `funnel-velocity.mjs` with an export contract that #1694 hasn't merged yet, and (b) `templates/benchmarks.yml` as a default source from #1693. Both guarded and tested-in-absence, so it's not a defect — it's sequencing: an API should be born in the PR where it's reviewed, not in its consumer. Given the block order this resolves itself: **once #1694/#1695 land, rebase here and the hooks stop being speculative** (they validate against merged code). If #1694 changes its exports during review, adjust here to match — the reviewed PR wins.

Also on the rebase: nothing else needed — CLAUDE.md/AGENTS/DATA_CONTRACT untouched by this PR, which made this the cleanest diff of the block.

(Web note: the oferta.md Block G source change is additive and contract-tests are green; I'm flagging the web maintainer in parallel — that thread is mine, not yours.)

**Comment by santifer** [comment]
Heads-up, @Scott-Emberson: this one has drifted into conflict with `main` and can't be merged as-is. Nothing wrong with the PR: the repo moved underneath it (we're landing ~10 a day right now), and GitHub is quiet about it, so people often don't find out for weeks.

I'm only pinging the PRs I actually want to land, and this is one of them: employer silence and repeat-poster history is real signal for a candidate deciding where to spend effort.

A rebase onto current `main` is all it needs. Ping me when it's green and I'll pick it up.

**Comment by santifer** [comment]
Status update rather than silence, @Scott-Emberson, since this has been open a while: it's green (2107 assertions, 0 failures on my run), `company-history.mjs` and its test are properly registered in SYSTEM_PATHS, and the coverage validator agrees.

What's holding it is not the code. It touches `modes/_shared.md`, which governs scoring across every evaluation the tool does, and by our own rule that file is Santiago's call rather than mine. It's in front of him with the analysis attached; I'm not going to merge around that even when the diff looks right to me, because "the maintainer merged a scoring change on a Sunday night" is exactly the kind of thing that rule exists to prevent.

The rest of it I'd take as-is. Employer-silence and repost facts are real signal for deciding where to spend effort, and it's the sort of thing only visible once someone has enough history to aggregate.

**Comment by santifer** [comment]
Nineteen days is too long for this to have sat, @Scott-Emberson, and I'm sorry. Santiago and I went through it properly this morning. **Most of it lands as-is; there's one piece I want to hold back, and I want you to see the reasoning rather than just the verdict, because it isn't a style preference.**

## What I checked, so you know this was read and not skimmed

Merged into current `main` it's green at **2293 assertions, 0 failures**, and the SYSTEM_PATHS coverage validator passes — both new top-level files are registered. `company-history.mjs` has zero write calls, so "read-only" is literally true and not just intended. Empty company keys land in `dataQuality.unjoinable` instead of being folded into some other company's card, which is the obvious way this kind of join goes wrong and you didn't do it. And the instruction to pass the company name as its own quoted argument, never spliced into a longer shell string, means you thought about company names containing quotes.

The part I most want to keep is the one that's easiest to get wrong: **`responsiveness` is explicitly non-scoring.** *"Must NOT alter the 1-5 score and must NOT alter the Assessment tier — those are driven exclusively by the `postingChurn` axis."* That distinction is the whole thing. A company ignoring you last March says nothing about whether *this* posting is real, and folding it into legitimacy would have turned a useful fact into a bias against employers who simply have bad ATS hygiene. You drew that line yourself and defended it in the mode text.

## What lands now

Everything except two hunks: the `modes/_shared.md` line and the Reposting Detection change in `modes/oferta.md`. That means `company-history.mjs`, its tests, the docs, the `followup-cadence.mjs` parser export, the runner wiring, and the whole prior-contact FYI go in. The genuinely new capability — joining the tracker, follow-ups and scan-history per company so a candidate can see their own history before spending an evening on a CV — arrives intact.

## What I'm holding, and the actual reason

The signal table in `_shared.md` currently reads like this. Look at the Source column:

| Signal | Source |
|---|---|
| Posting age | Page snapshot |
| Apply button active | Page snapshot |
| Tech specificity in JD | JD text |
| Requirements realism | JD text |
| Recent layoff news | WebSearch |
| **Reposting pattern** | **scan-history.tsv** |
| Salary transparency | JD text |
| Role-company fit | Qualitative |

**Every source is something the agent can look at.** A page, some text, a search result, a file. Not one of them requires that a program run successfully first. That isn't an accident of how it grew — it means a legitimacy assessment can always be completed with whatever is in front of you, and a degraded result is a *visible* gap rather than a silent one.

Your change would make Reposting pattern the first signal sourced from executing a script. The failure mode that introduces is specific: a user on an older checkout who hasn't run `update-system`, or a permissions problem, or any execution error, and the repost signal doesn't degrade — **it disappears, and Block G is scored without it.** No error surfaces, because from the agent's side "the card came back empty" and "there are no reposts" look identical. The user sees a legitimacy score with a missing leg and no way to know.

I've spent this week digging exactly that class of bug out of the scanner: a location filter where `"IN"` matched `"Indianapolis"` and quietly dropped real US postings; a Greenhouse path where the work model sat in `location.name` and the whole board was discarded; a sweep that hung on stalled response bodies with no error at all. Every one of them was correct-looking code whose failure was invisible. That's why this particular property of the table is worth defending even when the individual change is well built.

## And the part that I think actually settles it

What the change buys is company-key normalization: `postingChurn` re-keys clusters through `normalizeCompany`, so "Acme Inc." and "Acme" stop being two companies. **That's a real improvement and you're right that reading the TSV raw misses it.** But look at where the problem lives:

```
url	date	source	title	company	action
https://job-boards.greenhouse.io/airtable/...	2026-04-10	greenhouse-api	AI Agent Architect	Airtable	added
```

`scan-history.tsv` stores the **raw** company string and no normalized key. That's a *data* gap, not a *source* gap. Routing the signal through a script to fix it works, but it fixes name matching by changing what the scoring layer depends on, which is a much bigger lever than the problem needs.

Fix it in the file and everyone who reads it gets the benefit — the mode, the web, `detect-reposts.mjs`, anything future — with nothing depending on execution. A normalized company column is additive, so it doesn't break the scan-history contract the web parses. That's a small, self-contained PR, and I'd take it quickly. If you want it, it's yours; if you'd rather not, say so and I'll open it as a good first issue with your name on the reasoning.

## Being straight about the tradeoff

If I'm wrong, the cost is a few weeks of missed reposts at companies whose name is written two ways. Recoverable, and your later PR recovers it properly.

If I'm right and this goes in, a signal that helps someone decide whether a posting is a ghost can evaporate silently. Not recoverable, because nobody finds out.

That asymmetry is the whole argument. It isn't about your code, which is better than most of what I merge — it's about which lever the improvement is attached to.

Two smaller things while you're in there: CodeRabbit auto-paused this review ("branch under active development"), so `@coderabbitai review` would be worth triggering on 1,400 new lines. And if you disagree with any of the above, push back — you have 38 merges here and I'd rather be argued out of this than have you recut something you think is wrong.

**Comment by Scott-Emberson** [comment]
Agreed — pulling both hunks (the `_shared.md` Reposting-pattern row and the `oferta.md` Reposting Detection step). You're right, and it's the asymmetry that convinces me, not politeness: every other signal in that table completes with whatever's in front of the agent, and a script-sourced one fails to *absence* rather than to a *visible gap* — "empty card" and "no reposts" are indistinguishable from the agent's side, so Block G gets scored a leg short with nothing to flag it. That's the exact invisible-failure class you've been digging out of the scanner all week, and it's not worth introducing to fix a name-matching problem.

And the diagnosis lands: it's a data gap, not a source gap. `scan-history.tsv` carrying the raw company string with no normalized key is where "Acme Inc." and "Acme" split — fixing it in the file is additive, keeps the web parser contract intact, and every reader (mode, web, detect-reposts) gets it for free with nothing depending on execution.

**I'll take the follow-up** — normalized company column on `scan-history.tsv`, written at scan time through the same `normalizeCompany`, with `detect-reposts` preferring the column and falling back to normalizing the raw string for pre-existing rows. Separate small PR, coming shortly.

So this PR keeps company-history.mjs, its tests, the prior-contact FYI, the followup-cadence parser export and the runner wiring; the two reposting mode-hunks and the test-all assertion that asserted the script-sourced wiring come out. CodeRabbit review already re-triggered on the branch.

**Comment by santifer** [comment]
Merged, @Scott-Emberson. And thank you for the way you took the held-back part: you didn't just accept it, you restated the asymmetry better than I did ("empty card" and "no reposts" being indistinguishable from the agent's side is exactly the shape of it), and then you went and opened #2243 to fix the actual data gap. That's the follow-up done properly.

What lands here is the capability I wanted: someone can now see their own history with a company, follow-ups and scan-history joined, before spending an evening on a tailored CV. And the line you drew yourself on `responsiveness` being explicitly non-scoring is the part I'd have most regretted losing: a company with bad ATS hygiene isn't a company posting a ghost job, and folding those together would have quietly punished the wrong employers.

2292 assertions green, coverage validator clean, both new top-level files registered. #2243 is next up.
